runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java

```diff
@@ -86,13 +87,17 @@ public UnboundedReadFromBoundedSource(BoundedSource<T> source) {
 
   @Override
   public PCollection<T> expand(PBegin input) {
-    return input.getPipeline().apply(
+    PCollection<T> collection = input.getPipeline().apply(
         Read.from(new BoundedToUnboundedSourceAdapter<>(source)));
+
+    defaultOutputCoder = collection.getCoder();
+
+    return collection;
   }
 
   @Override
   protected Coder<T> getDefaultOutputCoder() {
-    return source.getDefaultOutputCoder();
+    return defaultOutputCoder;
```

Comment:
I didn't think of that this way, but after reading the code you mentioned it seems a far better idea. There are more PTransforms that need be modified the way you described. I'll also fix them in the next commit.